### ISSUE TITLE
Python 3 segfaults on exit interpreter

### ISSUE THREAD

**Issue by nim65s** [stack-of-tasks/eigenpy#35]
Hi,

As shown by #34, we have random segfaults on python3:
https://gepgitlab.laas.fr/gsaurel/eigenpy/-/jobs/2812

Here is a minimal working exemple triggering this segfault:

```py
import matrix
import numpy as np

m = np.reshape(range(64), (8, 8))
a = matrix.reflex(m, False)
print('end of script')
```

And here is what gdb finds:

```
(gdb) run python/test_mwe_segfault.py
Starting program: /usr/bin/python3 python/test_mwe_segfault.py
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff2f16700 (LWP 25047)]
[New Thread 0x7ffff0715700 (LWP 25048)]
[New Thread 0x7fffeff14700 (LWP 25049)]
[New Thread 0x7fffeb713700 (LWP 25050)]
[New Thread 0x7fffe8f12700 (LWP 25051)]
[New Thread 0x7fffe6711700 (LWP 25052)]
[New Thread 0x7fffe5f10700 (LWP 25053)]
end of script

Thread 1 "python3" received signal SIGSEGV, Segmentation fault.
0x00000000005a0ee8 in dict_dealloc.lto_priv.164 (mp=0x7ffff6955e48) at ../Objects/dictobject.c:1595
1595	../Objects/dictobject.c: Aucun fichier ou dossier de ce type.
(gdb) bt
#0  0x00000000005a0ee8 in dict_dealloc.lto_priv.164 (mp=0x7ffff6955e48) at ../Objects/dictobject.c:1595
#1  0x0000000000597903 in module_dealloc.lto_priv () at ../Objects/moduleobject.c:638
#2  0x00007ffff64e3d57 in eigenpy::PyMatrixType::~PyMatrixType() ()
   from /local/users/gsaurel/stack-of-tasks/eigenpy/libeigenpy.so
#3  0x00007ffff7829ff8 in __run_exit_handlers (status=0, listp=0x7ffff7bb45f8 <__exit_funcs>, 
    run_list_atexit=run_list_atexit@entry=true) at exit.c:82
#4  0x00007ffff782a045 in __GI_exit (status=<optimized out>) at exit.c:104
#5  0x00007ffff7810837 in __libc_start_main (main=0x4cfdd0 <main>, argc=2, argv=0x7fffffffd7a8, init=<optimized out>, 
    fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffd798) at ../csu/libc-start.c:325
#6  0x00000000005d6049 in _start ()
```

**Comment by jcarpent** [comment]
It appears that this bug is currently appearing on Python 3.4 which is the default version on Ubuntu 14.04.

**Comment by nim65s** [comment]
I can (sometimes) reproduce it on Python 3.5 in Ubuntu 16.04:

```
(gdb) run python/test_segfault.py 
Starting program: /usr/bin/python3 python/test_segfault.py
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
end of script
[Inferior 1 (process 4366) exited normally]
(gdb) run python/test_segfault.py 
Starting program: /usr/bin/python3 python/test_segfault.py
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
end of script

Program received signal SIGSEGV, Segmentation fault.
0x00000000005a0ee8 in ?? ()
(gdb) bt
#0  0x00000000005a0ee8 in ?? ()
#1  0x0000000000597903 in ?? ()
#2  0x00007ffff6840d67 in eigenpy::PyMatrixType::~PyMatrixType() ()
   from /root/robotpkg/math/py-eigenpy/work.7d0ef9700a43/eigenpy-1.4.3/libeigenpy.so
#3  0x00007ffff7829ff8 in __run_exit_handlers (status=0, listp=0x7ffff7bb45f8 <__exit_funcs>, 
    run_list_atexit=run_list_atexit@entry=true) at exit.c:82
#4  0x00007ffff782a045 in __GI_exit (status=<optimized out>) at exit.c:104
#5  0x00007ffff7810837 in __libc_start_main (main=0x4cfdd0 <main>, argc=2, argv=0x7fffffffe568, init=<optimized out>, 
    fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffe558) at ../csu/libc-start.c:325
#6  0x00000000005d6049 in _start ()
```

NB: To run gdb in docker, we have to turn off some security features. To get this, I had to use `docker run --cap-add=SYS_PTRACE --security-opt seccomp=unconfined --rm -it eur0c.laas.fr:5000/stack-of-tasks/eigenpy/py-eigenpy-py3:16.04`

**Comment by nim65s** [comment]
I can't reproduce it on Python 3.6 in Ubuntu 18.04

**Comment by jcarpent** [comment]
Which is also the case with OSX 10.13

**Comment by jcarpent** [comment]
I think we may close also this issue, and reopen it as soon as some new improvements have been made.

**Comment by nim65s** [comment]
BTW, this bug hits only when the interpreter exits… So everything works as intended, even if a segfault occurs when all the work is done, we can just ignore that.

**Comment by nim65s** [comment]
Following discussions in https://github.com/Gepetto/gepetto-viewer-corba/issues/85 I tried with python3.7 in a system entierely build with c++11, and the bug is still here.

**Comment by jmirabel** [comment]
A follow up on this, on 16.04, Python 3.5.2:
- numpy 1.11 (from ubuntu repo) I get a SEGV
- numpy 1.16.2 (from pip), I get no SEGV

**Comment by nim65s** [comment]
@jmirabel thanks for finding out this !

**Comment by olivier-stasse** [comment]
Does that mean that we have to package numpy v1.16.2 ?

**Comment by olivier-stasse** [comment]
I would prefer to avoid it.

**Comment by jmirabel** [comment]
No. #58 fixes the SEGV for numpy from the ubuntu repo.

**Comment by olivier-stasse** [comment]
Great !
Thank you very much.

**Comment by jcarpent** [comment]
Fixed by #58.
